### ISSUE TITLE
TFA_Optimization example not working

### ISSUE THREAD

**Issue by JacobHanimann** [BrentLab/TFA-evaluation#2]
Hi there, 
When running the TFA Inference example "python TFAinference.py -e hkMatrixE_optimize.csv -c hkSignedBinaryCS.csv -i 20 -v hkMatrixE_holdOut.csv -t Example" it returns the statement "Could not learn the activity matrix". I think it is because the line 228 "model.optimize()" in TFAinference.py always returns a False. Is there an easy fix for that? I would really like to try out your TFA inference approach, since I really like your TFA-evaluation method.
Thanks for the help!
Cheers, Jacob

**Comment by czma** [comment]
Hi Jacob,
Thanks for your interest! I'm not sure why the example isn't working for you, but hopefully it's something simple to troubleshoot:
1. Before you get the error, how many iterations are run? Or rather, is anything else printed besides that error message?
2.  Can you share the last few lines of logFiles/ExampleTFA.log ?
3. What version of gurobi are you using?

**Comment by JacobHanimann** [comment]
Thanks for the quick reply!

1. These are the things printed:
"
will constrain the signs/mode of regulation for CS values
Restricted license - for non-production use only - expires 2022-01-13
TFAinference.py:375: DeprecationWarning: Deprecated, pass a TempConstr or use Model.addLConstr
  Atemp = learnTFA(binaryTFA, startCS, expressionValues, False, args.fileTag+"TFA")
Could not learn the activity matrix here
"
I replaced addConstr() with addLConstr and the DeprecationWarning disappeared but didn't help otherwise.

2. The last log of  logFiles/ExampleTFA.log:
"
Gurobi 9.1.2 (mac64) logging started Tue Oct 26 18:24:21 2021

Changed value of parameter LogFile to logFiles/ExampleTFA.log
   Prev:   Default: 
Changed value of parameter Threads to 1
   Prev: 0  Min: 0  Max: 1024  Default: 0
Gurobi Optimizer version 9.1.2 build v9.1.2rc0 (mac64)
Thread count: 4 physical cores, 8 logical processors, using up to 1 threads
"

3. I'm using gurobipy 9.1.2 as stated in the log (and Python 3.8)

I tried to run the example on a second computer (also macbook) and got the same message: "Could not learn the activity matrix"

**Comment by czma** [comment]
For context, I'm running Gurobi 9.0.2 with Python 3.7.4, so that might be the reason for the error. That being said, I'd much rather get this code working on the new versions, rather than have everyone downgrade!
1. First suggestion has to do with the addContr function at lines 202 and 209. Try replacing:
`model.addConstr(constraint/constraintCounter, GRB.EQUAL, 1.0, "c"+str(i))`
with
`model.addConstr(constraint/constraintCounter==1.0, "c"+str(i))`
and replace:
`model.addConstr(varsMatrix[i][j], GRB.GREATER_EQUAL, 1.1*varsMatrix[i][-1])`
with 
`model.addConstr(varsMatrix[i][j] == 1.1*varsMatrix[i][-1])`
This is based on a comparison of the Gurobi documentation for constraints between the two versions:
[https://www.gurobi.com/documentation/9.0/refman/py_model_addconstr.html#pythonmethod:Model.addConstr](url)
[https://www.gurobi.com/documentation/9.1/refman/py_model_addconstr.html#pythonmethod:Model.addConstr](url)
2. If that doesn't work, maybe more information about the error can help. 
learnTFA() returns False at two points: line 230 and line 237. 
Right before line 237, you can print model.status to see what it is if not optimal, referencing here to interpret the number:
[https://www.gurobi.com/documentation/9.1/refman/optimization_status_codes.html](url)  
For line 230, try replacing:
```
  # Solve
  try:
  	model.optimize()
  except:
  	return False
```
with 
```
  # Solve
  try:
  	model.optimize()
  except GurobiError as e:
  	print(e.message)
        print(e.errno)
  	return False
```
and you can reference here to interpret the error number:
[https://www.gurobi.com/documentation/9.1/refman/error_codes.html#sec:ErrorCodes](url)

**Comment by JacobHanimann** [comment]
Alright I found out why it didn't work upon replacing line 230 to inspect the error. The output was: 

_"will constrain the signs/mode of regulation for CS values
Restricted license - for non-production use only - expires 2022-01-13
Model too large for size-limited license; visit https://www.gurobi.com/free-trial for a full license
10010
Could not learn the activity matrix here"_

I got the academic license now and that solves the issue for the size of the model. I thought the restricted license would be enough so my bad! Anyway, thanks a million for being so helpful!

Best, Jacob
